src/main/java/org/breedinginsight/services/TraitService.java

```diff
@@ -59,4 +83,107 @@
 
        return traitDAO.getTraitFull(programId, traitId);
     }
+
+    public List<Trait> createTraits(UUID programId, List<Trait> traits, AuthenticatedUser actingUser)
+            throws DoesNotExistException, ValidatorException {
+
+        Optional<Program> optionalProgram = programService.getById(programId);
+        if (!optionalProgram.isPresent()) {
+            throw new DoesNotExistException("Program does not exist");
+        }
+        Program program = optionalProgram.get();
+
+        Optional<User> optionalUser = userService.getById(actingUser.getId());
+        if (!optionalUser.isPresent()){
+            throw new InternalServerException("Could not find user in system");
+        }
+        User user = optionalUser.get();
+
+
+        // Get our ontology
+        Optional<ProgramOntology> programOntologyOptional = programOntologyService.getByProgramId(programId);
+        if (!programOntologyOptional.isPresent()){
+            throw new InternalServerException("Ontology does not exist for program");
+        }
+        ProgramOntology programOntology = programOntologyOptional.get();
+
+        ValidationErrors validationErrors = new ValidationErrors();
+
+        // Get our program observation levels
+        List<ProgramObservationLevel> programLevels = programObservationLevelService.getByProgramId(programId);
+        for (int i = 0; i < traits.size(); i++) {
+            Trait trait = traits.get(i);
+            if (trait.getProgramObservationLevel() != null){
+                List<ProgramObservationLevel> matchingLevels = programLevels.stream()
+                        .filter(programObservationLevel -> programObservationLevel.getName().equals(trait.getProgramObservationLevel().getName()))
+                        .collect(Collectors.toList());
+                if (matchingLevels.size() == 0) {
+                    ValidationError validationError = new ValidationError("programObservationLevel.name",
+                            "Program Observation Level does not exist.", HttpStatus.NOT_FOUND);
+                    validationErrors.addError(i, validationError);
+                } else {
+                    ProgramObservationLevel dbLevel = matchingLevels.get(0);
+                    trait.getProgramObservationLevel().setId(dbLevel.getId());
+                }
+            }
+        }
+
+        // Validate the traits
+        ValidationErrors requiredFieldErrors = TraitValidator.checkRequiredTraitFields(traits);
+        ValidationErrors dataConsistencyErrors = TraitValidator.checkTraitDataConsistency(traits);
+        ValidationErrors duplicateTraits = traitValidator.checkDuplicateTraitsExisting(traits);
+        ValidationErrors duplicateTraitsInFile = TraitValidator.checkDuplicateTraitsInFile(traits);
+        validationErrors.mergeAll(requiredFieldErrors, dataConsistencyErrors, duplicateTraits, duplicateTraitsInFile);
```

Comment:
Yep, that's a good idea. I'll make that change.

Reply:
Changed and pushed. I wasn't able to do this exactly the way described in your comment because of the program observation level check above this block of code. I did combine all of these validations into one method call though.